drivers/led/sct2024.c

```diff
@@ -0,0 +1,219 @@
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(sct2024, CONFIG_LED_LOG_LEVEL);
+
+#define SCT2024_LED_COUNT 16
+/* SCT2024 can be chained up to control more LEDs.
+ * The driver can be extended to support chaining.
+ */
+#define SCT2024_MAX_CHAIN_LENGTH 1
+
+struct sct2024_cfg {
+	const struct device *spi_dev;
+	struct spi_config spi_cfg;
+	const struct gpio_dt_spec la_pin;
+	const struct gpio_dt_spec oe_pin;
+	const struct led_info *leds_info;
+};
+
+struct sct2024_data {
+	uint16_t led_bitmap[SCT2024_MAX_CHAIN_LENGTH]; /* Bitmap to track LED states */
+};
+
+static const struct led_info *sct2024_get_led_info(const struct device *dev, uint32_t led)
+{
+	const struct sct2024_cfg *cfg = dev->config;
+
+	if (led >= SCT2024_LED_COUNT * SCT2024_MAX_CHAIN_LENGTH) {
+		return NULL;
+	}
+
+	return &cfg->leds_info[led];
+}
+
+static int sct2024_get_led_index(const struct device *dev, uint32_t led)
+{
+	const struct led_info *led_info = sct2024_get_led_info(dev, led);
+
+	if (led_info == NULL) {
+		return -EINVAL;
+	}
+
+	return led_info->index;
+}
+
+static int sct2024_spi_write(const struct device *dev)
+{
+	const struct sct2024_cfg *cfg = dev->config;
+	struct sct2024_data *data = dev->data;
+	uint8_t buffer[SCT2024_MAX_CHAIN_LENGTH * 2] = {0};
+
+	/* Feature proposal: When SCT2024_MAX_CHAIN_LENGTH is greater than 1, limit the number of
+	 * iterations based on the greatest led index for the chain - detect the actual chain length
+	 */
+	BUILD_ASSERT(SCT2024_MAX_CHAIN_LENGTH == 1,
+		     "Driver currently supports only a single SCT2024 device in the chain");
+	for (int i = 0; i < SCT2024_MAX_CHAIN_LENGTH; i++) {
+		buffer[i * 2] = (data->led_bitmap[i] >> 8) & 0xFF; /* Upper 8 bits */
+		buffer[i * 2 + 1] = data->led_bitmap[i] & 0xFF; /* Lower 8 bits */
+	}
+
+	const struct spi_buf tx_buf = {
+		.buf = buffer,
+		.len = sizeof(buffer),
+	};
+	const struct spi_buf_set tx = {
+		.buffers = &tx_buf,
+		.count = 1,
+	};
+
+	int ret = spi_write(cfg->spi_dev, &cfg->spi_cfg, &tx);
+
+	/* Toggle LA pin to latch data */
+	gpio_pin_set(cfg->la_pin.port, cfg->la_pin.pin, 1);
+	k_sleep(K_NSEC(20));
+	gpio_pin_set(cfg->la_pin.port, cfg->la_pin.pin, 0);
+
+	return ret;
+}
+
+static int sct2024_write(const struct device *dev)
+{
+	const struct sct2024_cfg *cfg = dev->config;
+	struct sct2024_data *data = dev->data;
+
+	if (cfg->oe_pin.port) {
+		bool all_leds_off = true;
+
+		for (int i = 0; i < SCT2024_MAX_CHAIN_LENGTH; i++) {
+			if (data->led_bitmap[i] != 0) {
+				all_leds_off = false;
+				break;
+			}
+		}
+
+		if (all_leds_off) {
+			gpio_pin_set(cfg->oe_pin.port, cfg->oe_pin.pin, 0);
+			return 0;
+		}
+	}
+
+	int ret = sct2024_spi_write(dev);
+
+	if (cfg->oe_pin.port) {
+		gpio_pin_set(cfg->oe_pin.port, cfg->oe_pin.pin, 1);
+	}
+
+	return ret;
+}
+static int sct2024_led_on(const struct device *dev, uint32_t led)
+{
+	struct sct2024_data *data = dev->data;
+	int led_index;
+
+	led_index = sct2024_get_led_index(dev, led);
```

Comment:
There is a confusion here between LED node and controller indexes.

See https://github.com/zephyrproject-rtos/zephyr/pull/98635#pullrequestreview-3404759367 for a full explanation.